docs/key-issues/cross_platform.md

```diff
@@ -0,0 +1,147 @@
+[Buildroot](https://buildroot.org/) are projects specifically focused on
+cross-compilation for Linux embedded systems. More general-purpose packaging
+ecosystems often have toolchains and supporting infrastructure to cross-compile packages for their own needs - see, e.g., info for
+[Void Linux](https://github.com/void-linux/void-packages#cross-compiling),
+[conda-forge](https://conda-forge.org/),
+[Debian](https://wiki.debian.org/CrossCompiling) and
+[Nix](https://nixos.org/guides/cross-compilation.html).
+
+[^1]:
+    The "build", "host" and "target" terminology for identifying which system
+    is which in a cross-compilation setup is not consistent across build
+    systems and packaging tools. Always carefully check whether "build" means
+    the machine on which the compilation is run and "host" the machine on which
+    the produced binaries will run - or vice versa.
+
+Tools like [crossenv](https://github.com/benfogle/crossenv) can be used to trick
+Python into performing cross-platform builds. These tools use path hacks and
+overrides of known sources of platform-specific details (like `distutils`) to
+provide a cross-compilation environment. However, these solutions tend to be
+somewhat fragile as they aren't first-class citizens of the Python ecosystem.
+
+[The BeeWare Project](https://beeware.org) also uses a version of these
+techniques. For both the platforms it supports, BeeWare provides a custom
+package index that contains pre-compiled binaries ([Android](https://chaquo.com/pypi-7.0/);
+[iOS](https://anaconda.org/beeware/repo)). These binaries are produced using a
+set of tooling ([Android](https://github.com/chaquo/chaquopy/tree/master/server/pypi);
+[iOS](https://github.com/freakboy3742/chaquopy/tree/iOS-support/server/pypi))
+that is analogous to the tools used by conda-forge to build binary artefacts.
+
+
+## Problems
+
+There is currently a gap in _communicating target platform details to the
+build system_. While a build system like autoconf or CMake may support
+cross-platform compilation, and a project may be able to cross-compile binary
+artefacts, invocation of a `pyproject.toml` build hook typically assumes that the
+platform running the build will be the platform that ultimately runs the Python
+code. As a result, `sys.platform`, or the various attributes of the `platform`
+and `sysconfig` modules can't be used as part of the build process.
+
+_Running Python code_ for the host (cross) platform is not possible (modulo
+using an emulator), but Python packages have not designed for this. to be
+avoided. For example, `numpy` and `pybind11` ship headers and have
+`get_include()` functions in their main namespaces to obtain the path to those
+headers. That is clearly a problem, which packages dependending on those
+headers have to work around (often done by patching those packages with
+hardcoded paths within a cross-compilation setup).
+
+`pip` provides limited support for installing binaries for a different platform
```

Comment:
@freakboy3742 is this actually a problem? It doesn't seem clear to me that it is. What is the use case here to install packages on a system where you know they won't run?

Reply:
Many packages that use NumPy's C API call `import numpy; numpy.get_include()` to find the location of NumPy's C headers. If you're cross-compiling, and the build platform is different to the NumPy package's platform, then `import numpy` will of course fail. This can be worked around by setting an internal NumPy variable to prevent it from importing its native libraries. But since most Python packages are not written with cross-compilation in mind, they need to be patched to set this variable.

Reply:
... or to avoid `get_include` altogether and use a hardcoded path instead.

Reply:
So the question is, how does patching the import tie into pip needing to install packages for the cross compile platform.

I think that depending on the package, you don't need this at all. e.g. for pybind11, it is arch-independent (header only) so you can install it any which way and do the `import pybind11; pybind11.get_include()` from the native python executable.

This is a problem for numpy, though, because it also has static libraries. So you do need the right numpy for this. In that case, you do need to be able to pip install the cross-built numpy.

You *also* need to patch it or hardcode the path, since you need its binary artifacts, but don't want to execute its code.

Reply:
I'm very aware of the issues with `get_include`, which is why I added it in the paragraph above. My question is not about that though, it's about this paragraph (_"`pip` provides limited support ..."_).

Of all the cross-building I've seen, that never once included using `pip` to install the non-native flavor of a build dependency. Is that something you need in BeeWare, and if so could you explain in a bit more detail how this goes wrong?

Reply:
Sure, cross compiling in package manager based ecosystems, such as conda or void-linux or yocto, invariably have their own way to install non-native flavors for packages where the native flavor simply will not cut it.

I would still assume in at least some cases though, that specialized tooling doesn't exist and it's simplest to build the tooling around `pip install`. But it's probably the minority of cases.

Reply:
`crossenv`'s docs actually do use `pip` to install both build and host packages: https://crossenv.readthedocs.io/en/latest/quickstart.html#make-the-cross-environment. But that seems to work (at least according to the docs). 

The next lines: _"however, these flags only work for the selection of pre-built binary artefacts, and are therefore constrained to the set of platform and ABI tags published by the author."_ hint at what's missing. But I'm missing some concrete scenario/details of where it breaks down. Maybe its more about wheel tags in general that `pip` does not understand?

Reply:
@rgommers The use case is mobile, web, and embedded systems. The critical difference is that these are systems where you *can't* (or, at the very least, can't without extreme difficulty) run the development environment on the device where the software will be used. For example, if you're building an iPhone app, you building the app on macOS. You can't run gcc (or any other binary tooling) on the device; and Apple will outright reject apps that try to contact an external (non-AppStore) repository to download additional code, so you can't run pip on the device either. That means pip needs to be executed on macOS, targeting iOS.

Reply:
That much is clear, but the next sentences imply that there is support for this via `--platform` & co. I just tried, and while the UX is pretty bad it took me only a couple of minutes to convince `pip` to install a Windows wheel on my macOS system with:
```
pip install scikit-image --platform=win_amd64 --no-deps --target . --only-binary :all: --implementation=cp
```
So if you replace `win_amd64` with whatever the right tag is for iOS, doesn't that work?

Reply:
It does - however, it only works if you enable `--only-binary`. In practice, we've found that we can't enable `--only-binary` because there are a non-trivial number of non-binary packages that are only available as source packages, not as noarch wheels. So, in order to invoke pip in a single pass (thereby getting a consistent dependency resolution pass), you have to allow source packages, which then opens the door to pip trying to compile numpy (et al) from source.

Admittedly, this is an edge case, and one that would go away if publishing noarch wheels was something that was required (and/or there was cross-compilation support in pip) - but it's a limitation that exists. Hence: "limited support".

Reply:
Well... there are 2 possible fixes:

1. Make everyone do the right thing, and fix all historical packages published in the world where wheels weren't required; or
2. Modify pip to deal with this edge case of packaging.

I can't see any world where (1) is the easier solution :-)

Reply:
(2) isn't an option, because there is unfortunately no way for `pip` to know if the package authors simply neglected to upload a `-none-any` wheel, or if there's compiled code in there.

There are also pure Python packages that don't have wheels because their whole purpose is to run Python code or provide system-dependent metadata at build time, e.g. `oldest-supported-numpy` and `cupy`.

Reply:
Thinking some more about how this actually plays out in practice. When you're dealing with cross-compiling a large amount of packages, your main concern is that you have to cross-compile every non-pure wheel. For something like iOS there is nothing on PyPI. So you're setting up your own wheelhouse anyway. At best, you can use PyPI as a fallback for `-none-any` wheels. So when those are missing, you treat the package like you do all packages which aren't pure Python, and put a `-none-any` wheel in your own wheelhouse. So this issue is then "for each missing `-none-any` wheel I have to produce that wheel and host it myself". Which is extra work, but probably a minor amount compared to the task of building wheels for non-pure packages.

A significant improvement here would be if `pip install --platform ...` gained the capability to build wheels from sdists. But that already falls under "standardize cross-compiling". The `--platform ...` specifiers would need to be passed down to the build backend, which is only possible after standization.

So I'm thinking of moving the main description of what `pip` can do to the "current status" section, and under "problems" leave only the "every missing `-none-any` wheel incurs some extra effort". WDYT?

Reply:
> A significant improvement here would be if `pip install --platform ...` gained the capability to build wheels from sdists.

Many of these sdists are actually pure-Python, or have a pure-Python fallback if their extension modules fail to compile. So for Android, BeeWare / Chaquopy patches its copy of pip to remove the block on installing sdists when `--platform` is supplied. Instead, we monkey-patch setuptools to immediately throw an exception when it attempts to execute a native compiler. This allows most pure-Python fallbacks to work correctly, while providing a more comprehensible error message if there is no fallback.

@freakboy3742 can confirm this, but I think on iOS we work a bit differently. We monkey-patch the standard library so that pip, running on macOS, believes it's actually running on iOS. That means the `--platform` option is not required, so building sdists is no problem as long as they don't attempt to execute a native compiler.

Either way, we don't need to host none-any wheels anywhere. If the package is able to produce a none-any wheel, then pip will build it automatically.

Reply:
Thanks. That sounds pragmatic, and similar to what `crossenv` does. Ideally such patching won't be needed anymore in a future that has more solid cross-compilation support. But that does rely on something being standardized.

Reply:
Confirming that yes, Briefcase on iOS is effectively doing "crossenv-lite"; and yes, the best-case endpoint is that externally hacked-on crossenv-style approaches aren't needed.